protocol/v1/qbft/instance/change_round.go

```diff
@@ -104,26 +105,22 @@ func (i *Instance) uponChangeRoundFullQuorum() pipelines.SignedMessagePipeline {
 				return
 			}
 
-			notPrepared, highest, e := i.HighestPrepared(signedMessage.Message.Round)
+			prepared, highest, e := i.HighestPrepared(i.State().GetRound())
 			if e != nil {
 				err = e
 				return
 			}
 
-			var proposalData *specqbft.ProposalData
+			proposalData := &specqbft.ProposalData{
+				RoundChangeJustification: roundChanges,
+				PrepareJustification:     highest.RoundChangeJustification,
+			}
 
-			if notPrepared {
-				proposalData = &specqbft.ProposalData{
-					Data:                     i.State().GetInputValue(),
-					RoundChangeJustification: i.containersMap[specqbft.RoundChangeMsgType].ReadOnlyMessagesByRound(i.State().GetRound()),
-				}
+			if prepared {
+				proposalData.Data = i.State().GetInputValue()
```

Comment:
by spec if highest is prepared, need to use it's value. if not use from state

`// Chose proposal value.
		// If justifiedRoundChangeMsg has no prepare justification chose state value
		// If justifiedRoundChangeMsg has prepare justification chose prepared value
		valueToPropose := instanceStartValue
		if highestRCData.Prepared() {
			valueToPropose = highestRCData.PreparedValue
		}
`

Reply:
qbft/round_change.go:45

Reply:
@nivBlox Thanks, I missed the inversion

Reply:
@nivBlox I wanted to handle prepared first and then not prepared, then reverted this change and forgot to put the exclamation sign back, thanks